roll/websocket.py

```diff
@@ -0,0 +1,141 @@
+        # might be part of the closing handshake (closing frame)
+        if self.websocket.state != 3:  # not closed
+            self.websocket.data_received(data)
+        else:
+            # The websocket is closed and we still get data for it
+            # This is an unexpected problem. Let's do nothing
+            # about it
+            pass
+
+    def write(self, *args):
+        # We are in websocket land
+        # We are not supposed to write outside the websocket.
+        # Maybe log ?
+        self.writer.close()
+
+
+class WebsocketHandler:
+
+    Protocol = WSProtocol
+    timeout = 5
+    max_size = 2 ** 20  # 1 megabytes
+    max_queue = 64
+    read_limit = 2 ** 16
+    write_limit = 2 ** 16
+
+    def __init__(self, handler, subprotocols: list=None, **kwargs):
+        self.handler = handler
+        if subprotocols:
+            subprotocols = frozenset(subprotocols)
+        self.subprotocols = subprotocols
+
+    def handshake(self, request):
+        """Websocket handshake, handled by `websockets`
+        """
+        headers = []
+
+        def get_header(k):
+            return request.headers.get(k.upper(), '')
+
+        def set_header(k, v):
+            headers.append((k, v))
+
+        try:
+            key = handshake.check_request(get_header)
+            handshake.build_response(set_header, key)
+        except InvalidHandshake:
+            raise RuntimeError('Invalid websocket request')
+
+        subprotocol = None
+        ws_protocol = get_header('Sec-Websocket-Protocol')
+        if self.subprotocols and ws_protocol:
+            # select a subprotocol
+            client_subprotocols = tuple(
+                (p.strip() for p in ws_protocol.split(',')))
+            for p in client_subprotocols:
+                if p in self.subprotocols:
+                    subprotocol = p
+                    set_header('Sec-Websocket-Protocol', subprotocol)
+                    break
+
+        # write the 101 response back to the client
+        # We will use a real Response in the future.
+        rv = b'HTTP/1.1 101 Switching Protocols\r\n'
+        for k, v in headers:
+            rv += k.encode('utf-8') + b': ' + v.encode('utf-8') + b'\r\n'
+        rv += b'\r\n'
+        request.transport.write(rv)
+
+        # Return the subprotocol agreed upon, if any
+        return subprotocol
+
+    def switch_protocol(self, request):
+        subprotocol = self.handshake(request)
+        websocket = WebSocketCommonProtocol(
+            timeout=self.timeout,
+            max_size=self.max_size,
+            max_queue=self.max_queue,
+            read_limit=self.read_limit,
+            write_limit=self.write_limit
+        )
+        websocket.subprotocol = subprotocol
+        websocket.connection_made(request.transport)
+        websocket.connection_open()
+        wsprotocol = self.Protocol(websocket)
+        request.transport.set_protocol(wsprotocol)
+        return wsprotocol, websocket
```

Comment:
It is only just a semantic thing but it could be used when overridden. When i get a "switch_protocol", i expect to get the new protocol back, even if I don't use it correctly. This can be removed obviously, if you don't share that feeling.

Reply:
Alright, let's keep it as is.

Reply:
This is no longer an issue, we no longer have 2 distinct protocols. The proxy protocol (wsprotocol) is useless and only necessary if we were an extension. Now, it doesn't make sense, it was removed and only "websocket" remains. I should have realized that earlier.